src/lib.rs

```diff
@@ -18,6 +19,30 @@ mod error;
 mod latlon;
 mod multipolygon;
 
+#[cfg(feature = "lazy_static")]
+lazy_static::lazy_static! {
+    /// The country boundaries data for the whole world under ODBl license.
+    pub static ref COUNTRY_BOUNDARIES: CountryBoundaries = {
+        let data = if cfg!(feature = "with_ODBL_licensed_OSM_data_high") {
```

Comment:
Wouldn't 
```
CountryBoundaries::default()
```
or something like this be more rusty or as rusty?

Reply:
Not really. The `::default` is the standard implementation of the empty constructor (see Default trait). In Rust, anything you call on a struct either acts on that struct (i.e. a member function), or creates that struct in some way (constructors). Technically you could add something unrelated too, but there is no point -- from a caller's perspective, all functions have a `a::b::c::d()` path, so there is little difference if `c` is a `struct` or a `mod`.  That's why you usually just make a pub fn that you usually would want to put in the root of your lib to be available.  `static` is a bit different here, but similar enough -- it is a preloaded singleton. No point in re-parsing it on every call (create new instances, etc).

You may want to change the name - up to you.

Reply:
Right, so it is not possible to only implement the Default trait only if `cfg!(feature = "with_ODBL_licensed_OSM_data*"`?

Because I would kind of expect as a user of the lib that the default contructor was available if I selected that I wanted default data to be used. The lazy static is I guess orthogonal to that, after all, the lazy static could just call 
`pub static ref COUNTRY_BOUNDARIES: CountryBoundaries = { CountryBoundaries::default() }` then.

Reply:
it is totally possible to implement a trait based on a condition, but that's not the point.  (you can add `#[cfg(feature="...")]` to `impl Default on CountryBoundaries...`)

The issue is that `default()` in Rust means **instantiation** of an object with default values - so every call to it would create a new instance. This is not what you want here - you do not want users to create multiple identical read-only copies of the same struct.  Instead, you need the singleton pattern - one global readonly copy of the struct pre-loaded with specific data.  And for that you should use a public static object.  The lazy init will simply ensure the static value is properly loaded initialized (which could be a fairly lengthy operation).

Reply:
P.S. to put it differently -- it is very much "anti-Rust" to implement an interface that can be easily used in a sub-optimal way.  Having a public "default()" will likely result in some users calling it on every point lookup, and then thinking your lib is extremely slow.

Reply:
I see. In Java (or generally), I'd say whether to use a class as a singleton or not would be up to the user, as different patterns for singletons exist: static variables, singleton objects or managed by a dependency injection framework.
`lazy-static` is a crate for singleton objects. But it is a crate, i.e. not in the standard library. This means that I'd force upon the user of this crate a certain implementation of a singleton pattern, while there may be others. With implementing the default trait, I wouldn't do that, but at the same time make it very easy for users to "wrap" it into their singleton pattern of choice.

Managing this via a dependency injection framework which I expect also exist for Rust would make sense for use cases where the scope in which the object is used would actually be not equal to the application scope. E.g. let's say (as you point out) the data structure is used for boundaries other than country boundaries for a certain lengthy calculation, and to conserve memory space, after that, the memory is cleared again. (After all, let's not forget that loading it into memory takes something like 40ms (on a SSD).

Reply:
That's a fair point - you don't need to dictate singleton pattern on the lib users. There is a better path - instead of the lazy init, you can add this to your lib:

```rust
pub static COUNTRY_BOUNDARIES_ODBL_DATA_LARGE = include_bytes!("../data/boundaries360x180.ser");
pub static COUNTRY_BOUNDARIES_ODBL_DATA_MID = include_bytes!("../data/boundaries180x90.ser");
pub static COUNTRY_BOUNDARIES_ODBL_DATA_SMALL = include_bytes!("../data/boundaries60x30.ser");
```

And it doesn't even have to be feature-gated (or could all be a single optional feature).  The data will be included as binary blobs directly, but the user would use these static values only if needed.  If not use, the compiler will remove these values when linking.  The user will decide how to implement singleton (or context) pattern, while fully realizing that each data parse will result in some performance impact.  Note that this method is actually nearly identical to reading data from a separate file from a disk - simply because OS will memmap the entire binary to memory, but will only load the segments needed by the program - so a large data segment could be purged/unused, and that memory is no different than disk cache.

Reply:
That's the fundamental functionality of the [linker](https://en.wikipedia.org/wiki/Linker_(computing)) -- it traverses the entire call tree of the program, and only keeps the symbols that are being used.  I am not sure if this is something you can turn off.  This happens because Rust (as well as C/C++) create a single binary that contains your code plus everything it uses.  This is different from the dynamically linked libs (.so / .dll) -- those are also linked, but they contain all functions your library **exports**, plus all the functions from other crates that are used by the exported functions.

Reply:
That's nice. This is actually a thing that is not done in Java, because Java has reflection and that means that you could call methods dynamically by name (and more weird stuff). It takes third party tools like ProGuard or R8 (for Android) to actually remove unused code and they assume that no reflection is used, i.e. may remove too much unless the configuration file specifies not to remove that module, this class, that method etc.

I haven't read into .so / .dll yet but I guess this is nothing I have to worry about for this crate(?)

So anyway, 

> pub static COUNTRY_BOUNDARIES_ODBL_DATA_LARGE = include_bytes!("../data/boundaries360x180.ser");
> pub static COUNTRY_BOUNDARIES_ODBL_DATA_MID = include_bytes!("../data/boundaries180x90.ser");
> pub static COUNTRY_BOUNDARIES_ODBL_DATA_SMALL = include_bytes!("../data/boundaries60x30.ser");

looks quite good then. They'd need to be referred to differently in the readme and... uhm well, it looks like it would be a different PR then. Okay for you if I do that in a new branch and you review it?

Reply:
sure, go for it